File: DSA/SakshamDwivedi_590013147/Day23/D23P2.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node* next;
};

int size = 0;
struct node* head = NULL;
void createNode(int);
void Print();
void RemoveDups();

int main(){
    createNode(1);
    createNode(2);
    createNode(2);
    createNode(2);
    createNode(3);
    createNode(4);
    createNode(4);
    createNode(4);
    createNode(4);
    createNode(4);
    createNode(5);
    Print();
    printf("\n");
    RemoveDups();
    Print();
}

void createNode(int data){
       struct node* Node = (struct node*)malloc(sizeof(struct node));
       Node -> data = data;
       Node -> next = NULL;
       if(head == NULL){
            head = Node;
            size++;
            return;
       }
     struct node* current = head;
     while(current -> next != NULL){
      current = current -> next;
     }
     current -> next = Node;
     size++;
}

void Print(){
    struct node* current = head;
    while(current != NULL){
        if(current -> next != NULL){
        printf("%d -> ", current -> data);}
        else{printf("%d", current -> data);}
        current = current -> next;
    }

}

void RemoveDups() {
    struct node *trav1 = head;
    struct node *trav2 = head;
    struct node *travfree = NULL;
    trav2 = trav2 -> next;
    while(trav2 != NULL){
        if(trav1 -> data == trav2 -> data){
            trav1 -> next = trav2 -> next;
            travfree = trav2;
            trav2 = trav2 -> next;
            free(travfree);
            size--;
        }
        else{
            trav2 = trav2 -> next;
            trav1 = trav1 -> next;
        }
    }

}

File: DSA/SakshamDwivedi_590013147/Day23/D23P2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

size = 0
head = None

def create_node(data):
    global head, size
    new_node = Node(data)
    if head is None:
        head = new_node
        size += 1
        return
    current = head
    while current.next:
        current = current.next
    current.next = new_node
    size += 1

def print_list():
    current = head
    while current:
        if current.next:
            print(current.data, " -> ", end="")
        else:
            print(current.data)
        current = current.next

def remove_dups():
    global head, size
    trav1 = head
    trav2 = head.next if head else None

    while trav2:
        if trav1.data == trav2.data:
            trav1.next = trav2.next
            trav2 = trav2.next
            size -= 1
        else:
            trav1 = trav1.next
            trav2 = trav2.next

create_node(1)
create_node(2)
create_node(2)
create_node(2)
create_node(3)
create_node(4)
create_node(4)
create_node(4)
create_node(4)
create_node(4)
create_node(5)

print("Before : ", end="")
print_list()

print("After : ", end="")
remove_dups()
print_list()
